File: 0000_simplest_database/cpp/simplest_database.cpp
#include "simplest_database.h"
#include <fstream>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <vector>

SimplestDatabase::SimplestDatabase(const std::string &filename)
    : filename(filename) {}

void SimplestDatabase::set(const std::string &key,
                           const nlohmann::json &json_dict) {
  std::ofstream file;
  file.open(filename, std::ios::app);
  if (!file.is_open()) {
    throw std::runtime_error("Unable to open file for writing");
  }
  file << key << "," << json_dict.dump() << "\n";
  file.close();
}

nlohmann::json SimplestDatabase::get(const std::string &key) {
  std::ifstream file(filename);
  if (!file.is_open()) {
    throw std::runtime_error("Unable to open file for reading");
  }

  std::vector<std::string> matching_rows;
  std::string line;
  while (getline(file, line)) {
    std::istringstream iss(line);
    std::string k, v;
    if (getline(iss, k, ',') && getline(iss, v)) {
      if (k == key) {
        matching_rows.push_back(v);
      }
    }
  }
  file.close();

  if (matching_rows.empty()) {
    return nullptr;
  } else {
    return nlohmann::json::parse(matching_rows.back());
  }
}

File: 0000_simplest_database/python/simplest_database.py
import json


class SimplestDatabase:
    def __init__(self, filename='database'):
        self.filename = filename

    async def set(self, key, json_dict):
        with open(self.filename, 'a') as f:
            f.write(f'{key},{json.dumps(json_dict)}\n')


    async def get(self, key):
        matching_rows = []
        with open(self.filename, 'r') as f:
            for line in f:
                k, v = line.strip().split(',')
                if k == key:
                    matching_rows.append(v)

        return json.loads(matching_rows[-1]) if matching_rows else None